text/000-rustdoc-jump-to-definition.md

```diff
@@ -0,0 +1,144 @@
+
+Basically, the source code pages' UI doesn't change.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+**The biggest concern is actually about "scope expansion"**. If we add this feature, why not adding others like a "commit source code viewer", or a "git blame source code viewer", or "go to this line in github/gitlab"?
+
+Unfortunately, on this part, the best answer I can provide is that we will need to put clear limits to what we don't want. I think we need to keep rustdoc focused on its target: documentation. We provide source code pages, so this feature allows to browse them more efficiently by using information we already have. Anything more advanced would be very likely out-of-scope.
+
+**Another concern is**: "why should we implement this in rustdoc? Aren't there already existing tools doing it?"
+
+A good answer to this is actually the following scenario: you are on <docs.rs> and you're looking at a crate documentation. You want to see how something is implemented and click on the `[src]` link. At this point, you encounter an item used in this page but not defined in this page. Do you want to clone the crate locally to check it out or go to github/gitlab to check it there or do you prefer having the links directly available?
+
+**Last concern is about the maintenance**.
+
+Any feature has a maintenance cost, this one included. It was suggested to put this feature in its own crate. The only part that can be extracted is the span map builder. But it would very likely be more of a problem than a solution considering that rustdoc API doesn't allow it easily.
+
+One more argument about this is that the feature actually requires not that much code. It is split in two parts:
+
+ * https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/render/span_map.rs (which is basically a visitor gathering `span`)
+ * https://github.com/rust-lang/rust/blob/master/src/librustdoc/html/highlight.rs#L725-L753 (for generating links)
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+[Github Code navigation feature](https://docs.github.com/en/repositories/working-with-files/using-files/navigating-code-on-github): Relying on external (and private) tools doesn't seem like a good idea. If they decide to shut it down, we can't do anything about it. In addition to that, not all projects are on github, so it excludes them. It also requires an internet connection.
+
+rust-analyzer: This is a public opensource project. However, the setup isn't super easy (but it's a minor issue) and when using it on the rust compiler source code, it's quite resource-heavy. Also, if you want to look at a crate source code on <docs.rs>, you'd need to clone it locally to be able to use rust-analyzer on it.
+
+# Prior art
+[prior-art]: #prior-art
+
+As mentionned above, it is actually quite common for documentation tools to have a source viewer. Implementations vary a bit between them:
```

Comment:
Looks like jump to definition only has prior art in a single one of these

Reply:
I didn't finish this section actually. ^^'